interval/window.go

```diff
@@ -0,0 +1,310 @@
+package interval
+
+import (
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/values"
+)
+
+// TODO(nathanielc): Make the epoch a parameter to the window
+// See https://github.com/influxdata/flux/issues/2093
+const epoch = values.Time(0)
+
+var epochYear, epochMonth int64
+
+func init() {
+	ts := epoch.Time()
+	y, m, _ := ts.Date()
+	epochYear = int64(y)
+	epochMonth = int64(m - 1)
+}
+
+// Window is a description of an infinte set of boundaries in time.
+type Window struct {
+	// The ith window start is expressed via this equation:
+	//   window_start_i = zero + every * i
+	//   window_stop_i = zero + every * i + period
+	every      values.Duration
+	period     values.Duration
+	zero       values.Time
+	zeroMonths int64
+}
+
+// NewWindow creates a window which can be used to determine the boundaries for a given point.
+// Window boundaries are defined to start at the epoch plus the offset.
+// Each subsequent window starts at a multiple of the every duration.
+// Each window's length is the start boundary plus the period.
+// Every must not be a mix of months and nanoseconds in order to preserve constant time bounds lookup.
+func NewWindow(every, period, offset values.Duration) (Window, error) {
+	zero := epoch.Add(offset)
+	w := Window{
+		every:      every,
+		period:     period,
+		zero:       zero,
+		zeroMonths: monthsSince(zero),
+	}
+	if err := w.isValid(); err != nil {
+		return Window{}, err
+	}
+	return w, nil
+}
+
+func (w Window) isValid() error {
+	if w.every.IsZero() {
+		return errors.New(codes.Invalid, "duration used as an interval cannot be zero")
+	}
+	if w.every.IsMixed() {
+		const docURL = "https://v2.docs.influxdata.com/v2.0/reference/flux/stdlib/built-in/transformations/window/#calendar-months-and-years"
+		return errors.New(codes.Invalid, "duration used as an interval cannot mix month and nanosecond units").
+			WithDocURL(docURL)
+	}
+	// TODO(nathanielc): what about negative every is that allowed?
```

Comment:
Are you asking why its _not_ allowed? Good question.

Reply:
My experience with this is that negative numbers as part of the window gets confusing very quickly. I think it can be ok for a negative period, but a negative every means you are getting the windows in the opposite direction and your incoming data needs to be processed differently. If every is negative, then I think the incoming data needs to be in descending order for certain algorithms to work. These are very easy to overlook and create correctness bugs.

This may be wrong. The windowing is already complicated enough that I'm just not sure what the answer to this is.